docs/src/writing_good_rules.md

```diff
@@ -103,16 +103,197 @@ In fact, any number of scalar arguments is supported, as is returning a tuple of
 See docstrings for the comprehensive usage instructions.
 ## Write tests
 
-In [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
-there are fairly decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
-Take a look at existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests and you should see how to do stuff.
+[ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
+provides tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+Take a look at the documentation or the existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests to see how to write the tests.
+It's easy!
 
 !!! warning
-    Use finite differencing to test derivatives.
     Don't use analytical derivations for derivatives in the tests.
     Those are what you use to define the rules, and so can not be confidently used in the test.
     If you misread/misunderstood them, then your tests/implementation will have the same mistake.
+    Use finite differencing methods instead, as they are based on the primal computation.
 
 ## CAS systems are your friends.
 
 It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).
+
+## Which functions need rules?
+
+In principle, a perfect AD system only needs rules for basic operations and can infer the
+rules for more complicated functions automatically.
+In practice, performance needs to be considered as well.
+
+Some functions use `ccall` internally for performance reasons, for example [`^`]
+(https://github.com/JuliaLang/julia/blob/380abc808ad8fbd72a7a57f6ac0f811b7903046d/base/math.jl#L922).
+These functions can not be differentiated through by AD systems, and need custom rules.
+
+Other functions can in principle be differentiated through by an AD system, but there is a
+mathematical insight that can dramatically improve the computation of the derivative.
+An example is numerical integration, where writing a rule removes the need to perform
+AD through numerical integration.
+
+Furthermore, AD systems make different trade-offs in performance due to their design.
+This means that a certain rule will help one AD system, but not improve (and also not harm)
+another.
+Below, we list some patterns relevant for the [`Zygote`](https://github.com/FluxML/Zygote.jl)
+AD system.
+
+### Patterns that need rules in [`Zygote`](https://github.com/FluxML/Zygote.jl)
+
+There are a few classes of functions that Zygote can not differentiate through.
+Custom rules will need to be written for these to make AD work.
+
+Other patterns can be AD'ed through, but they are slow.
+
+#### Functions which mutate arrays
+For example,
+```julia
+function addone!(array)
+    array .+= 1
+    return sum(array)
+end
+addone!_cr(array) = addone!(array)
```

Comment:
if we're keeping the `*_cr` functions, i think i'd move it down to where it's used, and add a comment (in addition to the text box below)

```suggestion
```